File: sdk/java/src/test/java/ai/aerol/microvm/RedirectSafetyTest.java
package ai.aerol.microvm;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.http.HttpClient;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import org.junit.jupiter.api.Test;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import ai.aerol.microvm.model.BuildImageOptions;
import ai.aerol.microvm.model.BuildImagePushOptions;
import ai.aerol.microvm.model.PushWasmModuleOptions;

/**
 * Cross-origin redirect hygiene: a 3xx must never carry Authorization or the
 * X-Registry-* credentials to another origin, and a 307/308 must never replay
 * a credential-bearing body cross-origin. Redirects are followed manually (up
 * to 5 hops) with those rules applied.
 */
class RedirectSafetyTest {

    @Test
    void crossOriginRedirectIsFollowedWithoutAuthorizationOrRegistryHeaders() throws Exception {
        ConcurrentLinkedQueue<Map<String, List<String>>> seen =
            new ConcurrentLinkedQueue<Map<String, List<String>>>();
        AtomicInteger targetHits = new AtomicInteger();

        HttpServer target = startServer(exchange -> {
            targetHits.incrementAndGet();
            seen.add(exchange.getRequestHeaders());
            writeJson(exchange, 200, "{\"module_ref\":\"oci://x\",\"digest\":\"d\",\"size_bytes\":1}");
        });
        HttpServer redirector = startServer(exchange -> {
            exchange.getResponseHeaders().add("Location", serverUrl(target) + "/v1/wasm-modules/push");
            exchange.sendResponseHeaders(302, -1);
        });

        try {
            MicroVMClient client = clientFor(redirector);
            client.pushWasmModule(new PushWasmModuleOptions()
                .setName("mod")
                .setTag("latest")
                .setModule(new byte[] {1, 2, 3})
                .setRegistryUsername("registry-user")
                .setRegistryToken("registry-token-secret"));

            assertEquals(1, targetHits.get(), "redirect should be followed exactly once");
            Map<String, List<String>> headers = seen.peek();
            assertFalse(headers.containsKey("Authorization"), "cross-origin redirect leaked Authorization");
            assertFalse(headers.containsKey("X-Registry-Token"), "cross-origin redirect leaked X-Registry-Token");
            assertFalse(headers.containsKey("X-Registry-Username"), "cross-origin redirect leaked X-Registry-Username");
        } finally {
            redirector.stop(0);
            target.stop(0);
        }
    }

    @Test
    void crossOrigin307RefusesToReplayCredentialBodies() throws Exception {
        String secret = "super-secret-push-password";
        List<String> bodies = Collections.synchronizedList(new ArrayList<String>());
        AtomicInteger targetHits = new AtomicInteger();

        HttpServer target = startServer(exchange -> {
            targetHits.incrementAndGet();
            bodies.add(new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8));
            writeJson(exchange, 200, "{\"image\":\"built:latest\"}");
        });
        HttpServer redirector = startServer(exchange -> {
            exchange.getResponseHeaders().add("Location", serverUrl(target) + "/v1/images/build");
            exchange.sendResponseHeaders(307, -1);
        });

        try {
            MicroVMClient client = clientFor(redirector);
            BuildImagePushOptions push = new BuildImagePushOptions();
            push.registry = "ghcr.io/acme/app";
            push.username = "acme";
            push.password = secret;

            boolean refused = false;
            try {
                client.buildImage(Image.base("alpine"), new BuildImageOptions().setPush(push));
            } catch (RuntimeException ex) {
                refused = true;
            }
            for (String body : bodies) {
                assertFalse(body.contains(secret), "cross-origin redirect replayed a body containing credentials");
            }
            assertTrue(refused || targetHits.get() == 0,
                "cross-origin 307 with a credential-bearing body was followed; want refusal");
        } finally {
            redirector.stop(0);
            target.stop(0);
        }
    }

    @Test
    void sameOriginRedirectKeepsAuthorization() throws Exception {
        List<String> auths = Collections.synchronizedList(new ArrayList<String>());
        AtomicInteger hits = new AtomicInteger();
        HttpServer server = startServer(exchange -> {
            hits.incrementAndGet();
            if ("/health".equals(exchange.getRequestURI().getPath())) {
                exchange.getResponseHeaders().add("Location", "/health2");
                exchange.sendResponseHeaders(302, -1);
                return;
            }
            auths.add(exchange.getRequestHeaders().getFirst("Authorization"));
            writeJson(exchange, 200, "{\"status\":\"ok\"}");
        });

        try {
            MicroVMClient client = clientFor(server);
            client.health();
            assertEquals(2, hits.get(), "same-origin redirect should be followed");
            assertEquals("Bearer pat-token", auths.get(0), "same-origin redirect dropped Authorization");
        } finally {
            server.stop(0);
        }
    }

    /**
     * A caller-supplied {@link HttpClient} whose policy follows redirects would
     * replay Authorization (and X-Registry-*) to the redirect target inside the
     * JDK, before the SDK's manual policy runs. The client must refuse it, so
     * the second origin receives nothing.
     */
    @Test
    void injectedRedirectFollowingHttpClientCannotReachTheRedirectTarget() throws Exception {
        ConcurrentLinkedQueue<Map<String, List<String>>> seen =
            new ConcurrentLinkedQueue<Map<String, List<String>>>();
        AtomicInteger targetHits = new AtomicInteger();

        HttpServer target = startServer(exchange -> {
            targetHits.incrementAndGet();
            seen.add(exchange.getRequestHeaders());
            writeJson(exchange, 200, "{\"status\":\"ok\"}");
        });
        HttpServer redirector = startServer(exchange -> {
            exchange.getResponseHeaders().add("Location", serverUrl(target) + "/health");
            exchange.sendResponseHeaders(302, -1);
        });

        try {
            MicroVMConfig config = new MicroVMConfig()
                .setApiUrl(serverUrl(redirector))
                .setPatToken("pat-token")
                .setHttpClient(HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL).build());
            try {
                MicroVMClient client = new MicroVMClient(config);
                client.health();
            } catch (MicroVMException expected) {
                // Fail closed: the redirect-following client was refused.
            }
            for (Map<String, List<String>> headers : seen) {
                assertNull(headers.get("Authorization"),
                    "injected redirect-following client leaked Authorization to the second origin");
            }
            assertEquals(0, targetHits.get(),
                "injected redirect-following client reached the redirect target");
        } finally {
            redirector.stop(0);
            target.stop(0);
        }
    }

    private static MicroVMClient clientFor(HttpServer server) {
        return new MicroVMClient(
            new MicroVMConfig().setApiUrl(serverUrl(server)).setPatToken("pat-token")
        );
    }

    private static HttpServer startServer(ThrowingHandler handler) throws IOException {
        HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        server.createContext("/", exchange -> {
            try {
                handler.handle(exchange);
            } finally {
                exchange.close();
            }
        });
        server.start();
        return server;
    }

    private static String serverUrl(HttpServer server) {
        return "http://127.0.0.1:" + server.getAddress().getPort();
    }

    private static void writeJson(HttpExchange exchange, int status, String body) throws IOException {
        byte[] payload = body.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().add("Content-Type", "application/json");
        exchange.sendResponseHeaders(status, payload.length);
        try (OutputStream out = exchange.getResponseBody()) {
            out.write(payload);
        }
    }

    private interface ThrowingHandler {
        void handle(HttpExchange exchange) throws IOException;
    }
}

File: sdk/go/internal/apiclient/redirect_safety_test.go
package apiclient

import (
	"context"
	"io"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"sync"
	"testing"

	"github.com/aerol-ai/microvm/pkg/models"
)

type recordedRequest struct {
	authorization string
	registryToken string
	registryUser  string
	body          string
	contentLength int64
}

// TestCrossOriginRedirectStripsRegistryCredentials pins the redirect policy:
// following a cross-origin 3xx must never forward Authorization or the
// X-Registry-* headers to the redirect target.
func TestCrossOriginRedirectStripsRegistryCredentials(t *testing.T) {
	var mu sync.Mutex
	var hits []recordedRequest

	target := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		mu.Lock()
		hits = append(hits, recordedRequest{
			authorization: r.Header.Get("Authorization"),
			registryToken: r.Header.Get("X-Registry-Token"),
			registryUser:  r.Header.Get("X-Registry-Username"),
			body:          string(body),
		})
		mu.Unlock()
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"module_ref":"oci://x/mod:latest","digest":"sha256:1","size_bytes":3}`))
	}))
	defer target.Close()

	redirector := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Redirect(w, r, target.URL+"/v1/wasm-modules/push", http.StatusFound)
	}))
	defer redirector.Close()

	client := NewClient(redirector.URL, ClientOptions{PATToken: "pat-token", HTTPClient: &http.Client{}})
	_, err := client.PushWasmModule(context.Background(), models.PushWasmModuleOptions{
		Name:             "mod",
		Tag:              "latest",
		Module:           []byte("wasm"),
		RegistryUsername: "registry-user",
		RegistryToken:    "registry-token-secret",
	})
	if err != nil {
		t.Fatalf("PushWasmModule() error = %v", err)
	}

	mu.Lock()
	defer mu.Unlock()
	if len(hits) != 1 {
		t.Fatalf("target hits = %d, want 1", len(hits))
	}
	hit := hits[0]
	if hit.registryToken != "" {
		t.Errorf("cross-origin redirect leaked X-Registry-Token: %q", hit.registryToken)
	}
	if hit.registryUser != "" {
		t.Errorf("cross-origin redirect leaked X-Registry-Username: %q", hit.registryUser)
	}
	if hit.authorization != "" {
		t.Errorf("cross-origin redirect leaked Authorization: %q", hit.authorization)
	}
}

// TestCrossOriginRedirectRefusesBodyReplay pins the second half of the policy:
// a 307/308 that would replay a credential-bearing body to a different origin
// must be refused, not followed.
func TestCrossOriginRedirectRefusesBodyReplay(t *testing.T) {
	const secret = "super-secret-push-password"

	var mu sync.Mutex
	var hitCount int
	var bodies []string

	target := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		mu.Lock()
		hitCount++
		bodies = append(bodies, string(body))
		mu.Unlock()
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"image":"built:latest"}`))
	}))
	defer target.Close()

	redirector := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Location", target.URL+"/v1/images/build")
		w.WriteHeader(http.StatusTemporaryRedirect)
	}))
	defer redirector.Close()

	client := NewClient(redirector.URL, ClientOptions{PATToken: "pat-token", HTTPClient: &http.Client{}})
	_, err := client.BuildImageWithPush(context.Background(), "FROM alpine", &BuildImagePushSpec{
		Registry: "ghcr.io/acme/app",
		Username: "acme",
		Password: secret,
	})

	mu.Lock()
	defer mu.Unlock()
	for i, body := range bodies {
		if strings.Contains(body, secret) {
			t.Errorf("cross-origin redirect replayed request body %d containing credentials", i)
		}
	}
	if hitCount > 0 && err == nil {
		t.Errorf("cross-origin 307 with a credential-bearing body was followed (hitCount=%d); want refusal", hitCount)
	}
	if hitCount == 0 && err == nil {
		t.Errorf("cross-origin 307 with a credential-bearing body neither followed nor refused")
	}
}

// TestDefaultPortIsNotCrossOrigin pins that a redirect which only spells out a
// default port (https://h/x -> https://h:443/x) is same-origin, so credentials
// are not stripped unnecessarily.
func TestDefaultPortIsNotCrossOrigin(t *testing.T) {
	cases := []struct {
		a, b string
		want bool
	}{
		{"https://h/x", "https://h:443/x", false},
		{"http://h/x", "http://h:80/x", false},
		{"https://h/x", "https://h:8443/x", true},
		{"https://h/x", "http://h/x", true},
		{"https://h:443/x", "https://other:443/x", true},
	}
	for _, tc := range cases {
		a, err := url.Parse(tc.a)
		if err != nil {
			t.Fatalf("url.Parse(%q) error = %v", tc.a, err)
		}
		b, err := url.Parse(tc.b)
		if err != nil {
			t.Fatalf("url.Parse(%q) error = %v", tc.b, err)
		}
		if got := isCrossOrigin(a, b); got != tc.want {
			t.Errorf("isCrossOrigin(%s, %s) = %v, want %v", tc.a, tc.b, got, tc.want)
		}
	}
}

// TestSameOriginRedirectKeepsAuthorization is the regression guard for the
// redirect policy: same-origin redirects keep working and keep auth.
func TestSameOriginRedirectKeepsAuthorization(t *testing.T) {
	var sawAuthOnSecondHit string
	var hits int
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		hits++
		if r.URL.Path == "/health" {
			http.Redirect(w, r, "/health2", http.StatusFound)
			return
		}
		sawAuthOnSecondHit = r.Header.Get("Authorization")
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"status":"ok"}`))
	}))
	defer server.Close()

	client := NewClient(server.URL, ClientOptions{PATToken: "pat-token", HTTPClient: server.Client()})
	if _, err := client.Health(context.Background()); err != nil {
		t.Fatalf("Health() error = %v", err)
	}
	if hits != 2 {
		t.Fatalf("hits = %d, want 2 (redirect should be followed same-origin)", hits)
	}
	if sawAuthOnSecondHit != "Bearer pat-token" {
		t.Errorf("same-origin redirect dropped Authorization: %q", sawAuthOnSecondHit)
	}
}
